File: app/src/main/java/com/example/tingximalaya/data/SubscriptionDao.java
package com.example.tingximalaya.data;


import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;


import com.example.tingximalaya.interfaces.ISubDao;

import com.example.tingximalaya.interfaces.ISubDaoCallBack;
import com.ximalaya.ting.android.opensdk.model.album.Album;
import com.ximalaya.ting.android.opensdk.model.album.Announcer;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class SubscriptionDao implements ISubDao {

    private static final SubscriptionDao ourInstance = new SubscriptionDao();
    private static final String TAG = "SubscriptionDao";
    private static XimalayaDbHelper mXimalayaDBHelper;
    private ISubDaoCallBack mCallback = null;
    private static Context mcontext = null;

    public static SubscriptionDao getInstance(Context context) {
        mcontext = context;
        if (mcontext != null) {

            mXimalayaDBHelper = new XimalayaDbHelper(mcontext);
        }

        return ourInstance;
    }

    private SubscriptionDao() {

    }


    @Override
    public void setCallBack(@NotNull ISubDaoCallBack callback) {
        this.mCallback = callback;

    }

    @Override
    public void addAlbum(Album album) {
        SQLiteDatabase db = null;
        boolean isAddSuccess = false;
        try {

            db = mXimalayaDBHelper.getWritableDatabase();
            db.beginTransaction();
            ContentValues contentValues = new ContentValues();
            //封装数据
            contentValues.put(Constants.SUB_COVER_URL, album.getCoverUrlLarge());
            contentValues.put(Constants.SUB_TITLE, album.getAlbumTitle());
            contentValues.put(Constants.SUB_DESCRIPTION, album.getAlbumIntro());
            contentValues.put(Constants.SUB_TRACKS_COUNT, album.getIncludeTrackCount());
            contentValues.put(Constants.SUB_PLAY_COUNT, album.getPlayCount());
            contentValues.put(Constants.SUB_AUTHOR_NAME, album.getAnnouncer().getNickname());
            contentValues.put(Constants.SUB_ALBUM_ID, album.getId());
            //插入数据
            db.insert(Constants.SUB_TB_NAME, null, contentValues);
            db.setTransactionSuccessful();
            isAddSuccess = true;
        } catch (Exception e) {
            e.printStackTrace();
            isAddSuccess = false;
        } finally {
            if (db != null) {
                db.endTransaction();
                db.close();
            }
            if (mCallback != null) {
                mCallback.addResult(isAddSuccess);
            }
        }
    }

    @Override
    public void delAlbum(Album album) {

        SQLiteDatabase db = null;
        boolean isDeleteSuccess = false;
        try {
            System.out.println("1111111111111111111111111111-------->删除");
            db = mXimalayaDBHelper.getWritableDatabase();
            db.beginTransaction();
            int delete = db.delete(Constants.SUB_TB_NAME, Constants.SUB_ALBUM_ID + "=?", new String[]{album.getId() + ""});
            db.setTransactionSuccessful();
            isDeleteSuccess = true;
        } catch (Exception e) {
            e.printStackTrace();
            isDeleteSuccess = false;
        } finally {
            if (db != null) {
                db.endTransaction();
                db.close();
            }
            if (mCallback != null) {
                mCallback.onDelResult(isDeleteSuccess);
            }
        }
    }

    @Override
    public void listAlbum() {
        SQLiteDatabase db = null;
        List<Album> result = new ArrayList<>();
        try {


            db = mXimalayaDBHelper.getReadableDatabase();
            db.beginTransaction();
            Cursor query = db.query(Constants.SUB_TB_NAME, null, null, null, null, null, "_id desc");
            //封装数据
            while (query.moveToNext()) {
                Album album = new Album();
                //封面图片
                String coverUrl = query.getString(query.getColumnIndex(Constants.SUB_COVER_URL));
                album.setCoverUrlLarge(coverUrl);
                //
                String title = query.getString(query.getColumnIndex(Constants.SUB_TITLE));
                album.setAlbumTitle(title);
                //
                String description = query.getString(query.getColumnIndex(Constants.SUB_DESCRIPTION));
                album.setAlbumIntro(description);
                //
                int tracksCount = query.getInt(query.getColumnIndex(Constants.SUB_TRACKS_COUNT));
                album.setIncludeTrackCount(tracksCount);
                //
                int playCount = query.getInt(query.getColumnIndex(Constants.SUB_PLAY_COUNT));
                album.setPlayCount(playCount);
                //
                int albumId = query.getInt(query.getColumnIndex(Constants.SUB_ALBUM_ID));
                album.setId(albumId);
                String authorName = query.getString(query.getColumnIndex(Constants.SUB_AUTHOR_NAME));
                Announcer announcer = new Announcer();
                announcer.setNickname(authorName);
                album.setAnnouncer(announcer);
                result.add(album);
            }
            query.close();
            db.setTransactionSuccessful();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (db != null) {
                db.endTransaction();
                db.close();
            }
            //把数据通知出去
            if (mCallback != null) {
                mCallback.onSubListLoaded(result);
            }
        }
    }


}

File: app/src/main/java/com/example/tingximalaya/data/SubscriptDao/SubscriptionDao.kt
package com.example.tingximalaya.data.SubscriptDao

import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import com.example.tingximalaya.base.BaseAplication
import com.example.tingximalaya.interfaces.ISubDao
import com.example.tingximalaya.interfaces.ISubDaoCallBack
import com.example.tingximalaya.utils.Constants
import com.ximalaya.ting.android.opensdk.model.album.Album
import com.ximalaya.ting.android.opensdk.model.album.Announcer
import java.lang.Exception


/**
 * @return:$
 * @since: 1.0.0
 * @Author:$
 * @Date: 2020/2/26$ 17:17$
 */
@SuppressLint("StaticFieldLeak")
object SubscriptionDao : ISubDao {


    private var mISbuCallBack: ISubDaoCallBack? = null


    private var mcontext: Context? = null

    private var dbhelper: XimalayaDbHelper? = null


    init {
        mcontext = BaseAplication().AppContext()!!
        if (mcontext != null) {
            dbhelper = XimalayaDbHelper(mcontext!!)
        }
    }


    override fun addAlbum(album: Album) {
        var db: SQLiteDatabase? = null
        try {
            db = dbhelper?.writableDatabase
            db?.beginTransaction()
            val contentValue by lazy { ContentValues() }
            //封装
            contentValue.put(Constants.SUB_COVER_URL, album.coverUrlLarge)
            contentValue.put(Constants.SUB_TITLE, album.albumTitle)
            contentValue.put(Constants.SUB_DESCRIPTION, album.albumIntro)
            contentValue.put(Constants.SUB_TRACKS_COUNT, album.includeTrackCount)
            contentValue.put(Constants.SUB_PLAY_COUNT, album.playCount)
            contentValue.put(Constants.SUB_AUTHOR_NAME, album.announcer.nickname)
            contentValue.put(Constants.SUB_ALBUM_ID, album.id)

            db?.insert(Constants.SUB_TB_NAME, null, contentValue)
            db?.setTransactionSuccessful()
            if (mISbuCallBack != null) {
                mISbuCallBack?.addResult(true)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            if (mISbuCallBack != null) {
                mISbuCallBack?.addResult(false)
            }
        } finally {
            db?.close()
            db?.endTransaction()
        }

    }

    override fun delAlbum(album: Album) {
        var db: SQLiteDatabase? = null
        try {
            db = dbhelper?.writableDatabase
            db?.beginTransaction()

            db?.delete(
                Constants.SUB_TB_NAME, Constants.SUB_ALBUM_ID + "?",
                (album.id.toString() as List<String>).toTypedArray()
            )


            db?.setTransactionSuccessful()

            if (mISbuCallBack != null) {
                mISbuCallBack?.addResult(true)
            }
        } catch (e: Exception) {
            if (mISbuCallBack != null) {
                mISbuCallBack?.addResult(false)
            }
            e.printStackTrace()
        } finally {
            db?.close()
            db?.endTransaction()
        }

    }

    override fun listAlbum() {
        var list: ArrayList<Album> = ArrayList()
        var db: SQLiteDatabase? = null
        try {
            db = dbhelper?.readableDatabase
            db?.beginTransaction()
            var query = db?.query(Constants.SUB_TB_NAME, null, null, null, null, null, null)

            while (query?.moveToNext()!!) {
                var album = Album()
                var url = query.getString(query.getColumnIndex(Constants.SUB_COVER_URL))
                album.coverUrlLarge = url
                var title = query.getString(query.getColumnIndex(Constants.SUB_TITLE))
                album.albumTitle = title
                var description = query.getString(query.getColumnIndex(Constants.SUB_DESCRIPTION))
                album.albumIntro = description
                var trackcount = query.getString(query.getColumnIndex(Constants.SUB_TRACKS_COUNT))
                album.includeTrackCount = trackcount.toLong()
                var playcount = query.getString(query.getColumnIndex(Constants.SUB_PLAY_COUNT))
                album.playCount = playcount.toLong()
                var nickname = query.getString(query.getColumnIndex(Constants.SUB_AUTHOR_NAME))
                var announcer = Announcer()
                announcer.nickname = nickname
                album.announcer = announcer
                var id = query.getString(query.getColumnIndex(Constants.SUB_ALBUM_ID))
                album.id = title.toLong()
                list.add(album)
            }
            query.close()
            db?.setTransactionSuccessful()
            if (mISbuCallBack != null) {
                mISbuCallBack?.onSubListLoaded(list)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            if (mISbuCallBack != null) {
                mISbuCallBack?.addResult(false)
            }
        } finally {
            db?.close()
            db?.endTransaction()
        }
    }

    override fun setCallBack(callback: ISubDaoCallBack) {
        this.mISbuCallBack = callback
    }


}
